src/OrleansProviders/Streams/Common/PooledCache/ObjectPool.cs

```diff
@@ -0,0 +1,44 @@
+﻿
+using System;
+using System.Collections.Generic;
+
+namespace Orleans.Providers.Streams.Common
+{
+    /// <summary>
+    /// Simple object pool that uses a stack to store available objects.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>
+    {
+        private const int DefaultPoolCapacity = 1 << 10; // 1k
+        private readonly Stack<T> pool;
+        private readonly Func<IObjectPool<T>, T> factoryFunc;
+
+        public ObjectPool(Func<IObjectPool<T>, T> factoryFunc, int initialCapacity = DefaultPoolCapacity)
+        {
+            if (factoryFunc == null)
+            {
+                throw new ArgumentNullException("factoryFunc");
+            }
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException("initialCapacity");
+            }
+            this.factoryFunc = factoryFunc;
+            pool = new Stack<T>(initialCapacity);
+        }
+
+        public virtual T Allocate()
+        {
+            return pool.Count != 0
+                ? pool.Pop()
+                : factoryFunc(this);
+        }
+
+        public void Free(T resource)
+        {
+            pool.Push(resource);
```

Comment:
could this be dangerous if we NEVER completely release a pooled resource? I thought the idea is to use individual PooledResources (which they contain several cached items) to reduce GC preassure, but here you are also never releasing the PooledResources themselves... basically by not allowing the GC to even collect the PooledResource... is that by design?